src/serac/physics/utilities/functional/domain_integral_cuda.cuh

```diff
@@ -130,12 +133,15 @@ __global__ void eval_cuda_element(const solution_type u, residual_type r, deriva
  * @see mfem::GeometricFactors
  * @param[in] num_elements The number of elements in the mesh
  * @param[in] qf The actual quadrature function, see @p lambda
+ * @param[inout] data The data for each quadrature point
  */
 
 template <Geometry g, typename test, typename trial, int Q, typename derivatives_type, typename lambda,
-          typename solution_type, typename residual_type, typename jacobian_type, typename position_type>
+          typename solution_type, typename residual_type, typename jacobian_type, typename position_type,
+          typename qpt_data_type = void>
 __global__ void eval_cuda_quadrature(const solution_type u, residual_type r, derivatives_type* derivatives_ptr,
-                                     jacobian_type J, position_type X, int num_elements, lambda qf)
+                                     jacobian_type J, position_type X, int num_elements, lambda qf,
+                                     QuadratureData<qpt_data_type>& data = dummy_qdata)
```

Comment:
do we want to pass QuadratureData by reference here? That is usually a recipe for trouble:

```
sam@provolone:~/code/cuda_sandbox$ cat global_in_kernel_by_value.cu 
#include <cstdio>

template < typename T >
struct QuadratureData{
  T value;
  __host__ __device__ T operator()(int i, int j) { 
    printf("calling QuadratureData<T>::operator() on device\n");
    printf("value: %f\n", value);
    return value; 
  }
};

QuadratureData<double> dummy;

__global__ void kernel(QuadratureData<double> data) {
  double value = data(0, 0);
}

int main() {
  dummy.value = 42.0;
  kernel<<<1,1>>>(dummy);
  cudaDeviceSynchronize();
}
sam@provolone:~/code/cuda_sandbox$ nvcc global_in_kernel_by_value.cu 
sam@provolone:~/code/cuda_sandbox$ cuda-memcheck ./a.out 
========= CUDA-MEMCHECK
calling QuadratureData<T>::operator() on device
value: 42.000000
========= ERROR SUMMARY: 0 errors
```

compared to

```
sam@provolone:~/code/cuda_sandbox$ cat global_in_kernel_by_reference.cu 
#include <cstdio>

template < typename T >
struct QuadratureData{
  T value;
  __host__ __device__ T operator()(int i, int j) { 
    printf("calling QuadratureData<T>::operator() on device\n");
    printf("value: %f\n", value);
    return value; 
  }
};

QuadratureData<double> dummy;

__global__ void kernel(QuadratureData<double> & data) {  // <-- emphasis mine
  double value = data(0, 0);
}

int main() {
  dummy.value = 42.0;
  kernel<<<1,1>>>(dummy);
  cudaDeviceSynchronize();
}
sam@provolone:~/code/cuda_sandbox$ nvcc global_in_kernel_by_reference.cu 
sam@provolone:~/code/cuda_sandbox$ cuda-memcheck ./a.out 
========= CUDA-MEMCHECK
========= Invalid __global__ read of size 8
=========     at 0x00000078 in kernel(QuadratureData<double>&)
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0x558193cd10b8 is out of bounds
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libcuda.so.1 [0x25428a]
=========     Host Frame:./a.out [0xc43b]
=========     Host Frame:./a.out [0x58ef0]
=========     Host Frame:./a.out [0x7f25]
=========     Host Frame:./a.out [0x7dcd]
=========     Host Frame:./a.out [0x7e19]
=========     Host Frame:./a.out [0x7c36]
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x270b3]
=========     Host Frame:./a.out [0x7aae]
=========
calling QuadratureData<T>::operator() on device
========= Program hit cudaErrorLaunchFailure (error 719) due to "unspecified launch failure" on CUDA API call to cudaDeviceSynchronize.
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libcuda.so.1 [0x355b43]
=========     Host Frame:./a.out [0x39b57]
=========     Host Frame:./a.out [0x7c3b]
=========     Host Frame:/usr/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x270b3]
=========     Host Frame:./a.out [0x7aae]
=========
========= ERROR SUMMARY: 2 errors
```

Reply:
<img width="1055" alt="thou_shalt_not_pass_by_reference" src="https://user-images.githubusercontent.com/61714427/132552506-648fc5de-432b-4952-befb-1c47c594999d.png">

Reply:
Good catch! I've made some changes that remove the by-reference parameters from the `__global__` functions but of course open to further suggestions/feedback